protocols/pnet/src/crypt_writer.rs

```diff
@@ -0,0 +1,146 @@
+        }
+    }
+
+    /// Gets a pinned mutable reference to the inner writer.
+    ///
+    /// It is inadvisable to directly write to the inner writer.
+    pub fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut W> {
+        self.project().inner
+    }
+}
+
+/// Write the contents of a Vec<u8> into an AsyncWrite.
+///
+/// The handling 0 byte progress and the Interrupted error was taken from BufWriter in async_std.
+///
+/// If this fn returns Ready(Ok(())), the buffer will be completely flushed.
+fn poll_flush_buf<W: AsyncWrite>(
+    inner: &mut Pin<&mut W>,
+    buf: &mut Vec<u8>,
+    cx: &mut Context<'_>,
+) -> Poll<io::Result<()>> {
+    let mut ret = Poll::Ready(Ok(()));
+    let mut written = 0;
+    let len = buf.len();
+    while written < len {
+        match inner.as_mut().poll_write(cx, &buf[written..]) {
+            Poll::Ready(Ok(n)) => {
+                if n > 0 {
+                    // we made progress, so try again
+                    written += n;
+                } else {
+                    // we got Ok but got no progress whatsoever, so bail out so we don't spin writing 0 bytes.
+                    ret = Poll::Ready(Err(io::Error::new(
+                        io::ErrorKind::WriteZero,
+                        "Failed to write buffered data",
+                    )));
+                    break;
+                }
+            }
+            Poll::Ready(Err(e)) => {
+                // Interrupted is the only error that we consider to be recoverable by trying again
+                if e.kind() != io::ErrorKind::Interrupted {
+                    // for any other error, don't try again
+                    ret = Poll::Ready(Err(e));
+                    break;
+                }
+            }
+            Poll::Pending => break,
+        }
+    }
+    if written > 0 {
+        buf.drain(..written);
+    }
+    ret
+}
+
+impl<W: AsyncWrite> AsyncWrite for CryptWriter<W> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut this = self.project();
+        // completely flush the buffer, returning pending if not possible
+        ready!(poll_flush_buf(&mut this.inner, this.buf, cx))?;
+        // if we get here, the buffer is empty
+        let res = Pin::new(&mut *this.buf).poll_write(cx, buf);
+        if let Poll::Ready(Ok(count)) = res {
+            this.cipher.apply_keystream(&mut this.buf[0..count]);
+            trace!("encrypted {} bytes", count);
+        };
+        // flush immediately afterwards, but if we get a pending we don't care
+        if let Poll::Ready(Err(e)) = poll_flush_buf(&mut this.inner, this.buf, cx) {
+            Poll::Ready(Err(e))
+        } else {
+            res
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let mut this = self.project();
+        ready!(poll_flush_buf(&mut this.inner, this.buf, cx))?;
```

Comment:
Wrong if `poll_flush_buf` doesn't guarantee an empty buffer.

Reply:
As mentioned above, I think that poll_flush_buf does guarantee that when it returns Poll::Ready(Ok(())), the buffer is indeed empty. If this is not the case, this line is broken.

But if poll_flush_buf returns Pending or an error, isn't it OK here to pass through that error / pending using the ready! macro and the ? suffix? If you return Pending, whoever calls this method will try again until that is no longer the case. And if you return an error, whoever calls this method is supposed to stop. At least that is how I interpreted the contract for the poll_xxx methods.